### ISSUE TITLE
Implementation about kernel activation

### ISSUE THREAD

**Issue by lifuguan** [ZwwWayne/K-Net#21]
Hello, 
        Sorry to disturb you. I'm trying to visualize the **kernels** (called `object_feats` in your code). It've been illustrated in your paper.
![image](https://user-images.githubusercontent.com/17793134/148923944-116c47af-1116-47ce-b4a9-b3c5f874d0dd.png)
Here is my code, which aims to save and add them on `kernels.npy` during the inference phrase.

```python
"""kernel_iter_update.py line:296"""
                results.append(single_result)
        from debugger import save_test_info
        save_test_info(img_metas, scores_per_img, masks_per_img, object_feats)
        return results


def save_test_info(img_metas:list, 
    cls_score:torch.Tensor, 
    scaled_mask_preds:torch.Tensor, obj_feats:torch.Tensor):
    ...
    # kernels
    if obj_feats is not None:
        kernels_old = np.load("work_dirs/tmp/kernels.npy")
        kernels_new = obj_feats.to('cpu').detach().numpy()
        kernels = kernels_new+kernels_old
        np.save("work_dirs/tmp/kernels.npy", kernels)
```
```python
"""after inference phrase"""
fig,a =  plt.subplots(10,10)
kernels_2dim = kernels.reshape((100,16,16))
for i in range(100):
    # a[int(i / 10)][i % 10].set_title(i)
    a[int(i / 10)][i % 10].set_xticks([])
    a[int(i / 10)][i % 10].set_yticks([])
    a[int(i / 10)][i % 10].imshow(kernels_2dim[i], cmap = plt.cm.hot_r)
plt.savefig('work_dirs/tmp/class_80_ins_2/kernel_2dim.png', bbox_inches='tight')
plt.show()
```
However, the result is completely different from your figures:
![image](https://user-images.githubusercontent.com/17793134/148924922-fb2b4846-c5a9-49c2-a005-c8879cd76489.png)

## It will be appreciated if anyone can show me the way to visualize **kernel** correctly.

**Comment by ZwwWayne** [comment]
You should not save kernels, but the masks. The paper says "... the average of mask activations of the 100 instance kernels over the 5000 images in the val split. " Thus, the mask predictions after sigmoid are saved and averaged, rather than the kernels.

**Comment by lifuguan** [comment]
Thank you. I've saved the `masks_per_img` which corresponds to **mask predictions after sigmoid** you mentioned above.
However, the results are still confusing:
![image](https://user-images.githubusercontent.com/17793134/154634398-7c768c5d-4e46-4f73-8381-5009e1ab0f30.png)

Here is my code:
```python
def save_test_info(img_metas:list, 
    cls_score:torch.Tensor, 
    scaled_mask_preds:torch.Tensor):
    # activation average mask
    bn = nn.BatchNorm2d(100, affine=False).cuda()
    sigmoid = nn.Sigmoid()
    scaled_mask_preds = sigmoid(bn(scaled_mask_preds.unsqueeze(0)))
    if scaled_mask_preds is not None:
        scaled_mask_preds = F.interpolate(scaled_mask_preds, size=(200, 300))
        activation_new = scaled_mask_preds.squeeze(0).to('cpu').detach().numpy()

        if os.path.exists("work_dirs/tmp/activation.npy") is False:
            np.save("work_dirs/tmp/activation.npy", activation_new)
        else:
            activation_old = np.load("work_dirs/tmp/activation.npy")
            activation = activation_new+activation_old
            np.save("work_dirs/tmp/activation.npy", activation)
```
